### ISSUE TITLE
Error while replacing identifier of a PropertyAccessExpression

### ISSUE THREAD

**Issue by gregjacobs** [dsherret/ts-simple-ast#365]
Hey @dsherret, long time no talk! How's it going?

I'm back doing ts transforms (yay!), but ran into an issue while trying to replace the object name (expression) in a PropertyAccessExpression.

I'm basically trying to convert `that.something` to `this.something`, but if the there are 4 chained PropertyAccessExpressions, I get an error. Try running this code:

```
import Project, { Identifier, PropertyAccessExpression, SyntaxKind, TypeGuards } from 'ts-simple-ast';

const project = new Project();
const sourceFile = project.createSourceFile( 'my-file.ts', `
	that.something.something2.something3;   // <-- too nested!
` );


const thatDotSomething: PropertyAccessExpression = sourceFile
	.getDescendantsOfKind( SyntaxKind.PropertyAccessExpression )
	.find( propAccess => TypeGuards.isIdentifier( propAccess.getExpression() ) )!;

const thatIdentifier = thatDotSomething.getExpression() as Identifier;
thatIdentifier.replaceWithText( 'this' );

console.log( sourceFile.getFullText() );
```

I get the error: `Error: Error replacing tree: Should not have more children left over.`

Oddly enough, when I remove one property and change `that.something.something2.something3` to `that.something.something2`, the code works fine.

Can you take a look into this? Thanks!

**Comment by cancerberoSgx** [comment]
I think is an issue indeed. First of all I suggest you use rename() instead o replaceWithText(). Is safer, wont make the nodes obsolete and also it will rename all the references to that symbol in the whole project (just like TypeScript rename feature). Nevertheless rename() has the same problem. 

Fortunately you can workaround it by renaming the Identifier definition node:

[play with the code](https://typescript-api-playground.glitch.me/#__editor_code__=import%20Project%2C%20%7B%20Identifier%2C%20PropertyAccessExpression%2C%20SyntaxKind%2C%20TypeGuards%2C%20Node%2C%20DeclarationNamedNode%20%7D%20from%20'ts-simple-ast'%0Aexport%20function%20main(code%3Astring%2C%20log%3A%20(msg%3Astring)%3D%3Evoid)%20%7B%0A%0Aconst%20project%20%3D%20new%20Project(%7BuseVirtualFileSystem%3A%20true%7D)%3B%0Aconst%20sourceFile%20%3D%20project.createSourceFile(%20'my-file2.ts'%2C%20%60%0Alet%20a%2C%20x%0Aa.b.c.d%20%3D%203%3B%20%20%20%2F%2F%20%3C--%20too%20nested!%0Ax%3D1%0A%60%20)%3B%0A%0Aconst%20aInExpression%20%3D%20sourceFile.getDescendants().find(d%20%3D%3E%20TypeGuards.isIdentifier(d)%20%26%26%20%0A%20%20TypeGuards.isExpression(d.getParent())%20%26%26%20d.getText()%3D%3D%3D'a')%20as%20Identifier%0A%0A%2F%2F%20renaming%20the%20id%20in%20the%20expression%20doesn't%20work%3A%20%0A%2F%2F%20aInExpression.rename('RENAMED1')%3B%20%0A%0A%2F%2F%20however%20renaming%20the%20definitino%20node%20does%3A%0A%2F%2F%20aInExpression.getDefinitionNodes().filter(canBeRenamed).forEach(n%20%3D%3E%20n.rename('RENAMED2'))%20%2F%2F%20n%20is%20VariableDeclaration%0A%0A%2F%2F%20also%20renaming%20the%20id%20of%20the%20variable%20declaration%20(and%20not%20the%20variable%20declaration%20itseof)%20works%3A%0AsourceFile.getDescendants().filter(d%20%3D%3E%20TypeGuards.isIdentifier(d)%20%26%26%20%0A%20%20TypeGuards.isVariableDeclaration(d.getParent())%20%26%26%20d.getText()%3D%3D%3D'a').%0A%20%20filter(canBeRenamed).forEach(n%3D%3En.rename('RENAMED3'))%3B%20%0A%0Alog(sourceFile.getFullText())%3B%0A%7D%0Afunction%20canBeRenamed(node%3A%20Node)%3A%20node%20is%20Node%20%26%20%7B%20rename%3A%20(name%3A%20string)%3D%3E%20Node%3B%20%7D%20%7B%0A%20%20return%20typeof%20(node%20as%20any).rename%20%3D%3D%3D%20'function'%0A%7D%0A__editor_input__=class%20A%20%7B%0A%20%20color%3A%20string%0A%20%20method%20(a%3A%20number%2C%20b%3A%20Date%5B%5D%5B%5D)%3A%20Promise%3Cvoid%3E%20%7B%0A%20%20%20%20return%20Promise.resolve()%0A%20%20%7D%0A%7D%0Aconst%20a%20%3D%20new%20A()%0A%20%20%20%20%20%20)
```ts
import Project, { Identifier, PropertyAccessExpression, SyntaxKind, TypeGuards, Node, DeclarationNamedNode } from 'ts-simple-ast'

function canBeRenamed(node: Node): node is Node & { rename: (name: string)=> Node; } {
  return typeof (node as any).rename === 'function'
}

const project = new Project({useVirtualFileSystem: true});
const sourceFile = project.createSourceFile( 'my-file2.ts', `
let a, x
a.b.c.d = 3;   // <-- too nested!
x=1
` );

const aInExpression = sourceFile.getDescendants().find(d => TypeGuards.isIdentifier(d) && 
  TypeGuards.isExpression(d.getParent()) && d.getText()==='a') as Identifier

// renaming the id in the expression doesn't work: 
// aInExpression.rename('RENAMED1'); 

// however renaming the definitino node does:
// aInExpression.getDefinitionNodes().filter(canBeRenamed).forEach(n => n.rename('RENAMED2')) // n is VariableDeclaration

// also renaming the id of the variable declaration (and not the variable declaration itseof) works:
sourceFile.getDescendants().filter(d => TypeGuards.isIdentifier(d) && 
  TypeGuards.isVariableDeclaration(d.getParent()) && d.getText()==='a').
  filter(canBeRenamed).forEach(n=>n.rename('RENAMED3')); 
log(sourceFile.getFullText());
```

**Comment by gregjacobs** [comment]
Interesting, I didn't realize there was a rename method. Thanks for that!

Although an interesting thing in this case might be that I'm "renaming" from an identifier node to a ThisKeyword node. Will that still work? (I haven't checked the code to see if there's a special case for the string "this")

And of course as you pointed out, the renaming still doesn't work on its own with this level of nested property accesses :/

**Comment by dsherret** [comment]
Hey @gregjacobs! I've been doing well thanks. I hope you have been too. Nice to see you back, but sorry that it's because of a bug!

Thanks for reporting this. I've fixed it now in 12.6.1. Just use the return value of `replaceWithText` to get the new `ThisKeyword` as the previous `Identifier` will be forgotten.

```ts
const thatIdentifier = ...;
const thisKeyword = thatIdentifier.replaceWithText("that");

thatIdentifier.getText(); // throws because node is forgotten
thisKeyword.getText();    // ok
```

**Comment by gregjacobs** [comment]
@dsherret Thanks for the quick fix! Definitely good to be back - I really love this stuff and I missed it :)

Very good to know about the return value from `replaceWithText()` - didn't realize that it did that. Many thanks, and keep up the great work!
